files/doom/config.org
#+TITLE: Doom Emacs Configuration
#+STARTUP: overview

* Basic Configuration
** Personal Information
Some personal information which is used by gpg and the likes.
#+begin_src elisp
(setq user-full-name "Benjamin"
      user-mail-address "b3nj4m1n@gmx.net")
#+end_src
** Variables
#+begin_src elisp
;; Default directory for org files
(setq org-directory "~/.local/share/org/")
;; Default directory for org-roam
(setq org-roam-directory "~/.local/share/org")
#+end_src
** Functions
Toggle between the 2 most recent buffers.
#+begin_src elisp
(defun switch-to-previous-buffer ()
  (interactive)
  (switch-to-buffer (other-buffer (current-buffer) 1)))
#+end_src
** Keybindings
#+begin_src elisp
;; Use super as meta key
(setq x-super-keysym 'meta)
;; Evil specific keybindings
(after! evil
  ;; Exit insert mode by hitting Control + c twice
  (define-key evil-insert-state-map (kbd "C-c C-c") 'evil-normal-state)
  (define-key evil-normal-state-map (kbd "C-c C-c") 'evil-normal-state)
  ;; Paste with Control + Shift + v in insert mode
  (define-key evil-insert-state-map (kbd "C-S-v") 'evil-paste-before)
  ;; Execute source code blocks (In org-mode) by pressing Control + c Control + r(un)
  (define-key evil-normal-state-map (kbd "C-c C-r") 'org-babel-execute-src-block)
  ;; Kill the current buffer and switch to the previous one
  (define-key evil-normal-state-map (kbd "<SPC> b k") 'kill-this-buffer)
  ;; Toggle between the two most recent buffers
  (define-key evil-normal-state-map (kbd "<SPC> t <SPC>") 'switch-to-previous-buffer)
  )
#+end_src
** Appearance
#+begin_src elisp
;; Doom theme to use
(setq doom-theme 'doom-dracula)
;; Auto reload theme shortly after startup
(run-with-idle-timer 0.1 nil (lambda () (add-hook 'doom-load-theme-hook 'theme-magic-from-emacs)))
;; Use relative line numbers
(setq display-line-numbers-type 'relative)
#+end_src
* Evil
#+begin_src elisp
;; When substituting using evil, do global substitution by default
(after! evil (setq evil-ex-substitute-global t))
;; Don't use system clipboard by default
(setq x-select-enable-clipboard nil)
#+end_src
* Org Mode Configuration
#+begin_src elisp
;; Open org files with all headings collapsed
(setq org-startup-folded t)
(after! org
  ;; Default directory used for org files
  (setq org-directory "~/.local/share/org/")
  ;; Directory searched by org-agenda
  (setq org-agenda-files '("~/.local/share/org/"))
  ;; Default file used for notes ("notes.org" in the default org-directory)
  (setq org-default-notes-file (expand-file-name "notes.org" org-directory))
  ;; Symbol to show for a collapsed section
  (setq org-ellipsis " ▼ ")
  ;; Insert a timestamp when marking a todo item as done
  (setq org-log-done 'time)
  ;; Hide the characters surround emphasised blocks of text; For example the ~ around monospaced text
  (setq org-hide-emphasis-markers t)
  ;; Directory for storing journal entrys
  (setq org-journal-dir "~/.local/share/org/journal/")
  ;; Prefix for inserted time in journal entry; This will make a new heading for every time entry in the journal entry of a given day
  (setq org-journal-time-prefix "* ")
  ;; Set the date in a journal file as the title of the file
  (setq org-journal-date-prefix "#+TITLE: ")
  ;; The format used for the inserted date title
  (setq org-journal-date-format "%A, %d %B %Y")
  ;; The filename format used for the journal files
  (setq org-journal-file-format "%Y-%m-%d.org")
  ;; All capture templates; Currently set to none
  (setq org-capture-templates '())
  ;; Enable inline tasks
  (require 'org-inlinetask)
  )

;; ;; Make sure that org-roam is loaded correctly
;; (after! org-roam
;;   :ensure t
;;   :hook
;;   (after-init . org-roam-mode)
;;   :custom
;;   (org-roam-directory "~/.local/share/org"))
  #+end_src
* Email
#+begin_src elisp
;; Use mbsync as a sync backend for notmuch
(setq +notmuch-sync-backend 'mbsync)
#+end_src
* Completion
#+begin_src elisp
;; How long after staying idle company should start showing suggestions
(setq company-idle-delay 0)
;; How many characters you must have minimum for company to start showing suggestions
(setq company-minimum-prefix-length 0)
;; Don't show numbers next to companys suggestions
(setq company-show-numbers nil)
;; When the last entry in the completion list is reached, wrap around to the first one
(setq company-selection-wrap-around t)
#+end_src
* Dired
Use peep dired for file previews
#+begin_src elisp
(evil-define-key 'normal peep-dired-mode-map (kbd "<SPC>") 'peep-dired-scroll-page-down
  (kbd "C-<SPC>") 'peep-dired-scroll-page-up
  (kbd "<backspace>") 'peep-dired-scroll-page-up
  (kbd "j") 'peep-dired-next-file
  (kbd "k") 'peep-dired-prev-file)
(add-hook 'peep-dired-hook 'evil-normalize-keymaps)
#+end_src
* Spellchecking
#+begin_src elisp
;; ;; Spellcheck config
;; (with-eval-after-load "ispell"
;;   ;; Configure `LANG`, otherwise ispell.el cannot find a 'default dictionary' even though multiple dictionaries will be configured in next line.
;;   (setenv "LANG" "en_GB")
;;   ;; Backend program to be used by ispell
;;   (setq ispell-program-name "hunspell")
;;   ;; Default dictionary to be used by ispell
;;   (setq ispell-dictionary "en_GB")
;;   ;; ispell-set-spellchecker-params has to be called before ispell-hunspell-add-multi-dic will work
;;   (ispell-set-spellchecker-params)
;;   ;; Configure multiple dictionarys with hunspell
;;   (ispell-hunspell-add-multi-dic "de_DE,en_GB")
;;   ;; For saving words to the personal dictionary, don't infer it from the locale, otherwise it would save to ~/.hunspell_de_DE.
;;   (setq ispell-personal-dictionary "~/.local/share/dictionary/hunspell_personal"))

;; ;; Enable flyspell everywhere
;; (add-hook 'text-mode-hook 'flyspell-mode)
;; (add-hook 'prog-mode-hook 'flyspell-prog-mode)

;; ;; The personal dictionary file has to exist, otherwise hunspell will silently not use it.
;; (unless (file-exists-p ispell-personal-dictionary)
;;   (write-region "" nil ispell-personal-dictionary nil 0))
#+end_src

* Other
** Edit textareas in browser using emacs
#+begin_src elisp
(require 'atomic-chrome)
(atomic-chrome-start-server)
#+end_src


files/lemonblocks/config.txt
0,3,NULL,/usr/bin/lbscripts/powermenu.sh,l, ,NULL,50fa7b,NULL,NULL,alacritty &,NULL,NULL,%{+u}%{U%c2-hex},%{U-}%{-u}
1,4,NULL,/usr/bin/lbscripts/bspwm.sh,l, ,NULL,NULL,NULL,NULL,NULL,NULL,NULL,[,]
1,5,NULL,/usr/bin/lbscripts/mpd.sh,l, ,NULL,NULL,mpc toggle; pkill lemonblocks -5,NULL,NULL,mpc volume +1,mpc volume -1,%{+u}%{U%scd-hex},%{U-}%{-u}
1,15,NULL,xtitle -f %s,c, ,NULL,NULL,NULL,NULL,NULL,NULL,NULL,%{+u}%{U%c5-hex},%{U-}%{-u}
0,6,NULL,/usr/bin/lbscripts/volume.sh,r, ,NULL,8649e1,amixer set Master toggle; pkill lemonblocks -6,NULL,NULL,amixer -q sset Master 1%+; pkill lemonblocks -6,amixer -q sset Master 1%-; pkill lemonblocks -6,%{+u}%{U%prm-hex},%{U-}%{-u}
0,8,NULL,/usr/bin/lbscripts/free-space.sh,r, ,NULL,00ffff,NULL,NULL,NULL,NULL,NULL,%{+u}%{U%scd-hex},%{U-}%{-u}
0,9,NULL,/usr/bin/lbscripts/layout.sh,r, ,NULL,8649e1,NULL,NULL,NULL,NULL,NULL,%{+u}%{U%prm-hex},%{U-}%{-u}
20,11,NULL,/usr/bin/lbscripts/load-average.sh,r, ,NULL,00ffff,NULL,NULL,NULL,NULL,NULL,%{+u}%{U%scd-hex},%{U-}%{-u}
60,12,NULL,/usr/bin/lbscripts/uptime.sh,r, ,NULL,8649e1,NULL,NULL,NULL,NULL,NULL,%{+u}%{U%prm-hex},%{U-}%{-u}
300,13,NULL,/usr/bin/lbscripts/network.sh,r, ,NULL,00ffff,NULL,NULL,NULL,NULL,NULL,%{+u}%{U%scd-hex},%{U-}%{-u}
30,14,NULL,/usr/bin/lbscripts/date.sh,r, ,NULL,8649e1,NULL,NULL,NULL,NULL,NULL,%{+u}%{U%prm-hex},%{U-}%{-u}


files/legacy/doom/config.el


;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Benjamin"
      user-mail-address "b3nj4m1n@gmx.net")

;; Exit insert mode
(define-key evil-insert-state-map (kbd "C-c C-c") 'evil-normal-state)
(define-key evil-normal-state-map (kbd "C-c C-c") 'evil-normal-state)
;; Paste with Control + Shift + v in insert mode
(define-key evil-insert-state-map (kbd "C-S-v") 'evil-paste-before)


;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-dracula)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/.local/share/org/")

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

;; Use super as meta key
(setq x-super-keysym 'meta)


;; Configure org
(after! org
  (setq org-directory "~/.local/share/org/")
  (require 'org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  (setq org-agenda-files '("~/.local/share/org/"))
  (setq org-default-notes-file (expand-file-name "notes.org" org-directory))
  (setq org-ellipsis " ▼ ")
  (setq org-log-done 'time)
  (setq org-hide-emphasis-markers t)
  (setq org-journal-dir "~/.local/share/org/")
  (setq org-journal-date-format "%A, %d %B %Y")
  (setq org-journal-file-format "%Y-%m-%d.org")
  )


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

;; Spellcheck config
(with-eval-after-load "ispell"
  ;; Configure `LANG`, otherwise ispell.el cannot find a 'default
  ;; dictionary' even though multiple dictionaries will be configured
  ;; in next line.
  (setenv "LANG" "en_GB")
  (setq ispell-program-name "hunspell")
  ;; Configure German, Swiss German, and two variants of English.
  (setq ispell-dictionary "de_DE_neu,en_GB")
  ;; ispell-set-spellchecker-params has to be called
  ;; before ispell-hunspell-add-multi-dic will work
  (ispell-set-spellchecker-params)
  (ispell-hunspell-add-multi-dic "de_DE_neu,en_GB")
  ;; For saving words to the personal dictionary, don't infer it from
  ;; the locale, otherwise it would save to ~/.hunspell_de_DE.
  (setq ispell-personal-dictionary "~/.local/share/dictionary/hunspell_personal"))
;; Enable flyspell everywhere
(add-hook 'text-mode-hook 'flyspell-mode)
(add-hook 'prog-mode-hook 'flyspell-prog-mode)

;; The personal dictionary file has to exist, otherwise hunspell will
;; silently not use it.
                                        ; (unless (file-exists-p ispell-personal-dictionary)
                                        ;   (write-region "" nil ispell-personal-dictionary nil 0))


;; Completion config
(after! company
  (setq company-idle-delay 0.1
        company-minimum-prefix-length 2)
  (setq company-show-numbers t)
  (add-hook 'evil-normal-state-entry-hook #'company-abort)) ;; make aborting less annoying.

